### PR TITLE
improve docs for std::hint::unreachable_unchecked()

### LINKED ISSUE #59549: Confusing wording in documentation for std::hint::unreachable_unchecked()
The documentation for [`std::hint::unreachable_unchecked()`](https://doc.rust-lang.org/std/hint/fn.unreachable_unchecked.html) mentions two ways that it can affect a program:

1. It can enable the compiler to make optimizations based on the assumption that the function will never be executed.
2. If the function actually is executed, its behavior is undefined.

It then goes on to mention the `unreachable!()` macro, saying

> The `unreachable!()` macro is the safe counterpart of this function, which will panic instead when executed.

This statement could be read as addressing the 2nd effect, but not the 1st. In other words, it could be interpreted as saying "the compiler will assume that `unreachable!()` will never be executed, but in the event that it is it will cause a panic." It might be worth rewording this statement to make it clear that neither (1) nor (2) applies to the safe version.

### PR DESCRIPTION
**Pull request by Reconcyl**
Fixes #59549

### CODE REVIEW

File: src/libcore/hint.rs [hunk 1]
```diff
@@ -21,9 +21,8 @@ use intrinsics;
 /// difficult-to-debug problems.
 ///
 /// Use this function only when you can prove that the code will never call it.
-///
-/// The [`unreachable!()`] macro is the safe counterpart of this function, which
-/// will panic instead when executed.
+/// Otherwise, consider using the [`unreachable!()`] macro, which does not allow
```
Comment (steveklabnik) [inline src/libcore/hint.rs]: This is good, but can you remove the `()s`, both here and in the link target? I know they were there previously, but the style is not to include them when referring to the names.

Comment (Reconcyl) [reply src/libcore/hint.rs]: I added another commit to address this.

**Review by Dylan-DPC-zz — APPROVED** [review]

**Pull request merged**